### ISSUE TITLE
portage: unsupported layout

### ISSUE THREAD

**Issue by gmt** [lxc/lxd#6655]
* Distribution: gentoo
 * lxd mostly from git
 * Distribution version: ~amd64 w/ openrc init (rc_cgroup_mode="unified")

# Issue description

I cooked up an lxd based on a recent lxd git master (20191219), which mostly worked to get unified cgroups support working on my host.

However, I kept getting unexplained reversions to the old cgroups v1 behavior, with init wedging in guests.  Sometimes it would be fine for hours but, once it wedged I had to reboot to get things working again.

Eventually, and after not inconsiderable frustration, debugging, analysis, etc., I figured out this was caused by (of all things) portage!

# etiology: when/where/why/wtf

If you enable portage's "cgroup" feature, it will automatically make a mount like so\*:

```
# mount | grep cg
none on /sys/fs/cgroup type cgroup2 (rw,nosuid,nodev,noexec,relatime)
tmpfs on /sys/fs/cgroup/portage type cgroup (rw,nosuid,nodev,noexec,...
```

Eventually\*\* I figured out this would interact with the code at the end of lxd/cgroup/init.go, which handles automagic detection of cgroups "layout" aka version:
https://github.com/lxc/lxd/blob/17616633f8b5cee1922bf16c65230cb0416f01a3/lxd/cgroup/init.go#L379-L385

The presence of the "cgroup" mount causes lxd to decide it's on a hybrid-cgroup host, seek the cgroup-v1 hierarchy at hard-coded location /sys/fs/cgroup/systemd, and face-plant if not present.

Kind of a portage bug, but maybe something ought to be done about it in lxd anyhow?  Surely, portage is not the only program to make such cgroup v1 style mounts on unified cgroup systems (even if I can't think of any).  Possible ameliorations include:

- Provide a run-time cgroups layout frob, and rely on automagic detection only if the frob is not frobbed.
- Advise unified cgroups users to add cgroup_no_v1=all to their kernel command line, effectively making such issues impossible.
- Retool the detection automagic to have init-system-specific detections.  For example, under systemd, lxd could look at the systemd version and kernel cmdline, to decide which layout should apply.  On openrc, lxd could look at the rc_cgroups_mode setting in rc.conf, if present.  Seems like an awful lot of work, though... after all, this is just transitional code, even if the transition period is expected to be pretty long :confounded:
- Provide a build-time cgroups layout override.  A bad option, mostly (although I have used it, locally, as a work-around and it did do the trick).
- Make the hybrid mode more flexible, i.e., if lxd is going to treat such discovered cgroups-v1 mounts as important, perhaps their discovered path(s) should be used, instead of the hard-coded "systemd" path.

<hr />
* <sup>I presume (without having investigated) that in portage, the cgroup feature has not yet been ported to support cgroups v2.  Which probably means it's a noop with automatic reversion to legacy behavior.  So this mounted hierarchy is harmless cruft, so long as the relevant controllers are cg2-delegated.  Deactivating cgroup in FEATURES would probably have solved the problem in my particular case.</sup>

** <sup>read: with great suffering... had to learn stuff like what is "GOPATH" variable etc.  I suppose you may not be terribly impressed with my pain :unamused: </sup>

**Comment by stgraber** [comment]
@brauner can you make lxd handle this setup? Assuming it's not also going to confuse the hell out of liblxc.

**Comment by gmt** [comment]
> @brauner can you make lxd handle this setup? Assuming it's not also going to confuse the hell out of liblxc.

It doesn't, so long as I set ~~~lxc.mount.auto = proc:mixed sys:mixed cgroup:rw:force~~~ in lxc's common.conf.

**Comment by stgraber** [comment]
we'd have to make sure it doesn't get confused when actual limits are being set too

The default config doesn't result in any `lxc.cgroup2` config options, so that's a large part of the liblxc cgroup2 handling logic not being touched with a default container.

**Comment by brauner** [comment]
> @brauner can you make lxd handle this setup?

This is simply a bug in portage. Such a cgroup layout is not supported in any way and won't be. If we go done this route we'll be back at detecting cgroup mounts on random mountpoints anywhere on the system and we're back in 20xx.
I'm fixing this in so far as I'll have lxd look at /sys/fs/cgroup and if that's a cgroup2 mount we'll stop looking for any additional shenanigans such as the stuff that portage is doing. Note, I'm only fixing this since that's the correct behavior for us anyway and this is the same what liblxc is doing. But that's it. We can't special case random cgroup layouts unfortunately.

Assuming it's not also going to confuse the hell out of liblxc.

liblxc deals with this because it stops looking for any other cgroup mounts if it detected that /sys/fs/cgroup is a cgroup2 mount.

**Comment by gmt** [comment]
> > @brauner can you make lxd handle this setup?
> 
> This is simply a bug in portage. 

I think so too.  Portage has a lot of code, a long history and not a lot of developers; this could have easily been written 10 years ago and just received minimal bug-fixing since then.

I plan to see if I can figure out a way to fix portage and either send patches or bring it up on the ml.

> Such a cgroup layout is not supported in any way and won't be. If we go done this route we'll be back at detecting cgroup mounts on random mountpoints anywhere on the system and we're back in 20xx.

...  yeah, no clue about that.  I never really learned how to manipulate the v1 hierarchy, myself. Probably because lxc did everything I needed without ever having to look under the hood.

I'm guessing it seemed like I was asking for something complicated, but...

> I'm fixing this in so far as I'll have lxd look at /sys/fs/cgroup and if that's a cgroup2 mount we'll stop looking for any additional shenanigans such as the stuff that portage is doing.

... ^^ that's more like what I had in mind.

> But that's it. We can't special case random cgroup layouts unfortunately.

OK.  Obviously I've walked unknowingly into some kind of minefield... pretty sure I'm not looking for any feature that might be likely to give anyone PTSD, if that helps  :stuck_out_tongue_winking_eye:

tbc, my motivation that brought me here was trying to get cg v1 out of my life, once and for all, now that lxd master looked like it would work with unified.  Whatever stupid thing portage is doing is not important to me, I just was only hoping to ensure people in similar spots would not have to go through what I did, to get to the bottom of it.
